doc/source/tune/tutorials/tune_get_data_in_and_out.md

```diff
@@ -0,0 +1,516 @@
+
+```{contents}
+    :local:
+    :backlinks: none
+```
+
+Let's start by defining a simple Trainable function. We'll be expanding this function with different functionality as we go.
+
+
+```python
+import random
+import time
+import pandas as pd
+
+
+def training_function(config):
+    # For now, we have nothing here.
+    data = None
+    model = {"hyperparameter_a": None, "hyperparameter_b": None}
+    epochs = 0
+
+    # Simulate training & evaluation - we obtain back a "metric" and a "trained_model".
+    for epoch in range(epochs):
+        # Simulate doing something expensive.
+        time.sleep(1)
+        metric = (0.1 + model["hyperparameter_a"] * epoch / 100) ** (
+            -1
+        ) + model["hyperparameter_b"] * 0.1 * data["A"].sum()
+        trained_model = {"state": model, "epoch": epoch}
+```
+
+Our `training_function` function requires a pandas DataFrame, a model with some hyperparameters and the number of epochs to train the model for as inputs. The hyperparameters of the model impact the metric returned, and in each epoch (iteration of training), the `trained_model` state is changed.
+
+We will run hyperparameter optimization using the [Tuner API](tune-run-ref).
+
+
+```python
+from ray.tune import Tuner
+from ray import tune
+
+tuner = Tuner(training_function, tune_config=tune.TuneConfig(num_samples=4))
+```
+
+## Getting data in
+
+First order of business is to provide the inputs for the Trainable. We can broadly separate them into two categories - variables and constants.
+
+Variables are the parameters we want to tune. They will be different for every [Trial](tune_60_seconds_trials). For example, those may be the learning rate and batch size for a neural network, number of trees and the maximum depth for a random forest, or the data partition if you are using Tune as an execution engine for batch training.
+
+Constants are the parameters that are the same for every Trial. Those can be the number of epochs, model hyperparameters we want to set but not tune, the dataset and so on. Often, the constants will be quite large (eg. the dataset or the model).
+
+### Search space
+
+```{note}
+TL;DR - use the `param_space` argument to specify small, serializable constants and variables.
+```
+
+The first way of passing inputs into Trainables is the [*search space*](tune-key-concepts-search-spaces) (it may also be called *parameter space* or *config*). In the Trainable itself, it maps to the `config` dict passed in as an argument to the function. You define the search space using the `param_space` argument of the `Tuner`. The search space is a dict and may be composed of [*distributions*](<tune-sample-docs>), which will sample a different value for each Trial, or out of constant values. The search space may be composed out of nested dictionaries, and those in turn can have distributions as well.
```

Comment:
```suggestion
The first way of passing inputs into Trainables is the [*search space*](tune-key-concepts-search-spaces) (it may also be called *parameter space* or *config*). In the Trainable itself, it maps to the `config` dict passed in as an argument to the function. You define the search space using the `param_space` argument of the `Tuner`. The search space is a dict and may be composed of [*distributions*](<tune-sample-docs>), which will sample a different value for each Trial, or of constant values. The search space may be composed of nested dictionaries, and those in turn can have distributions as well.
```